### ISSUE TITLE
runtime error: slice bounds out of range

### ISSUE THREAD

**Issue by lastzero** [dsoprea/go-jpeg-image-structure#12]
Didn't want to ruin your weekend, but we've got more :)

This iPhone 5s JPEG causes the parser to fail with `runtime error: slice bounds out of range [:12279] with capacity 12278`

https://github.com/photoprism/photoprism/raw/develop/internal/meta/testdata/out-of-range-500.jpg

**Comment by dsoprea** [comment]
It's nothing if not a therapeutic side-task.

Looks like a glitchy image. Apparently the thumbnail is supposed to have one more byte than the size of the JPEG segment will accommodate. The JPEG segment is 12278 bytes long but the thumbnail is located at offset 10597 and is 1682 bytes long. This means that its last byte is _reportedly_ at offset 12278 whereas the last offset is actually 12277. Hence the overflow.

I'm assuming that you skip the metadata extraction when there is an error, which is the correct course.

**Comment by dsoprea** [comment]
I don't mind a fair amount of forgiveness, but if there's an actual data allocation or structural error within the file, I think there should be a genuine issue.

That said, this is just a thumbnail issue and would not normally arise without specifically asking for a thumbnail. So, if you have an error extracting the thumbnail, just skip it. If there arises an issue parsing specific tags (FYI), then you can also reimplement using the visitor approach (as opposed to a flat dump of all of the tags) and then skip that tags that can not successfully be extracted. So, in other words, this is inherently a high-level problem and there does exist contingencies in handling this type of problem while _enumerating through the tags_.

**Comment by lastzero** [comment]
Do you suggest extracting raw exif data and passing it to go-exif is wrong?

```go
jpegMp := jpegstructure.NewJpegMediaParser()

sl, err := jpegMp.ParseFile(fileName)

if err != nil {
  return err
}

_, rawExif, err = sl.Exif()
```

Should we rather extract individual fields using the reader interface you mentioned? Note that `exif.SearchFileAndExtractExif()` works for this file. Only the jpeg parser returns an error, so it's mainly a performance issue and maybe slight inconsistency.

**Comment by dsoprea** [comment]
The only reason that it works is because the brute-force method extracts the EXIF blob from its signature to the end of the file (because there's no size defined in the EXIF header). Therefore, that thumbnail tag is not up against a physical boundary of a slice like it was when we do a proper extraction of the EXIF blob from the JPEG data prior to trying to parse it (which is not to imply that the thumbnail is always the last thing in the EXIF data; it just is in this case).

TLDR: The thumbnail data is likely corrupt. Even if you can extract it without an error when doing a brute force extraction of the EXIF blob, the size of the thumbnail data was incorrect and the extraction is likely including data that did not belong to it.

**Comment by lastzero** [comment]
While you may be technically correct, I'd recommend a less disturbing error message (which could also point to a parser bug).

Looking for GPS tags after extracting Exif headers using the "brute-force" method sometimes errors with "EOF", guess that's a similar issue.

Our users will see this:

<img width="878" alt="Screenshot 2020-07-21 at 23 55 09" src="https://user-images.githubusercontent.com/301686/88111150-b5513400-cbad-11ea-8e52-de66ca1af2c6.png">
<img width="364" alt="Screenshot 2020-07-21 at 23 53 49" src="https://user-images.githubusercontent.com/301686/88111155-b71af780-cbad-11ea-8c45-819e29e42fc3.png">

**Comment by dsoprea** [comment]
The problem is inherently that the structure is broken, the result will be undefined, and the error message will always be disturbing. We can't properly handle it when the issue is merely caused by garbage data. The error could be anything and anywhere.

I'd also recommend you don't just flip over to a brute-force search. It's just a redundant search trying to find an explanation that doesn't end in a parse error **due to** garbage. If it failed the first time, there's probably a very good reason for it and if you're looking for some parse that doesn't end in failure, you'll almost certainly get a false positive instead.

**Comment by lastzero** [comment]
In fact brute-force worked as we're mostly happy with date, location and maybe camera / lens, although we take more IF available. The JPEG itself looks good and broken.

**Comment by dsoprea** [comment]
There's still a gap in my understanding about why you're running into thumbnail issues (where the only issue I observed was). Are you actually requesting the thumbnail?

**Comment by lastzero** [comment]
No, not at all: We call `Exif()` and pass the raw data to go-exif. I've only upgraded to v3 and added support for additional file formats (HEIF & TIFF) compared to your original code.

**Comment by dsoprea** [comment]
Ah. Because we're now estimating how much actual EXIF data there is (advisory only, to support both curiosity and debugging whenever either comes up) by keeping track of the furthest offsets being accessed, we'll seek through the thumbnail data if present. We'll no longer allow [terminal] failures because of it:

https://github.com/dsoprea/go-exif/commit/33ee3a8313dabf9054611087d8314d9b4057c883

**Comment by lastzero** [comment]
Updated: https://github.com/photoprism/photoprism/commit/c16e3741ee6e4ce01baa879ff64ea154bd96f917

May this "fix" the `EOF` error or are there additional code changes needed on our side to see any effect?

**Comment by dsoprea** [comment]
Sorry. I didn't update this go.mod . Done.

963817537a6ca888350ff760b3496cbb24efade5

**Comment by dsoprea** [comment]
How was the integration to go-exif/v3? It should've been fairly easy.

**Comment by lastzero** [comment]
See https://github.com/photoprism/photoprism/commit/716aa082d8b6ac548074928ade37f112621b20b7

Since we're wasting performance by reading everything, it was easy after looking at your cli command example. Would not have been easy by just looking at our existing code and reading error messages.
